rust/examples/l2cap_bridge.rs

```diff
@@ -0,0 +1,647 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Rust version of the Python `l2cap_bridge.py` found under the `apps` folder.
+
+use bumble::wrapper::{
+    device::Device,
+    l2cap::LeConnectionOrientedChannel,
+    logging::{bumble_env_logging_level, py_logging_basic_config},
+    transport::Transport,
+};
+use clap::Parser as _;
+use owo_colors::OwoColorize;
+use pyo3::{PyObject, PyResult, Python};
+use std::{future::Future, path, sync::Arc};
+use anyhow::anyhow;
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    net::tcp::{OwnedReadHalf, OwnedWriteHalf},
+    sync::{mpsc::Receiver, Mutex},
+};
+
+#[pyo3_asyncio::tokio::main]
+async fn main() -> PyResult<()> {
+    env_logger::builder()
+        .filter_level(log::LevelFilter::Info)
+        .init();
+
+    py_logging_basic_config(bumble_env_logging_level("WARNING"))?;
+
+    let cli = Cli::parse();
+
+    println!("<<< connecting to HCI...");
+    let transport = Transport::open(cli.hci_transport).await?;
+    println!("<<< connected");
+
+    let mut device = Device::from_config_file_with_hci(
+        &cli.device_config,
+        transport.source()?,
+        transport.sink()?,
+    )?;
+
+    device.power_on().await?;
+
+    match cli.subcommand {
+        Subcommand::Server { tcp_host, tcp_port } => {
+            let args = server_bridge::Args {
+                psm: cli.psm,
+                max_credits: cli.l2cap_coc_max_credits,
+                mtu: cli.l2cap_coc_mtu,
+                mps: cli.l2cap_coc_mps,
+                tcp_host,
+                tcp_port,
+            };
+
+            server_bridge::start(&args, &mut device).await?
+        }
+        Subcommand::Client {
+            bluetooth_address,
+            tcp_host,
+            tcp_port,
+        } => {
+            let args = client_bridge::Args {
+                psm: cli.psm,
+                max_credits: cli.l2cap_coc_max_credits,
+                mtu: cli.l2cap_coc_mtu,
+                mps: cli.l2cap_coc_mps,
+                bluetooth_address,
+                tcp_host,
+                tcp_port,
+            };
+
+            client_bridge::start(&args, &mut device).await?
+        }
+    };
+
+    // wait until user kills the process
+    tokio::signal::ctrl_c().await?;
+
+    Ok(())
+}
+
+/// L2CAP CoC server bridge: waits for a peer to connect an L2CAP CoC channel
+/// on a specified PSM. When the connection is made, the bridge connects a TCP
+/// socket to a remote host and bridges the data in both directions, with flow
+/// control.
+/// When the L2CAP CoC channel is closed, the bridge disconnects the TCP socket
+/// and waits for a new L2CAP CoC channel to be connected.
+/// When the TCP connection is closed by the TCP server, XXXX
+mod server_bridge {
```

Comment:
I think these are big enough to be their own files (straightforward once it's part of the top level CLI)

Reply:
Agreed.

Reply:
Done. Have a look and lmk if there is anything you would have structured differently.